src/common/libevent.hpp

```diff
@@ -463,6 +466,150 @@ class Bufferevent : public NonCopyable, public NonMovable {
     }
 };
 
+class EvdnsBase {
+  public:
+    // Add a reference to Var<EventBase> base to make sure it is referenced
+    Var<EventBase> evbase = nullptr;
+    evdns_base *dns_base = nullptr;
+
+    EvdnsBase() {}
+    ~EvdnsBase() {}
+
+    template <decltype(evdns_base_new) construct = ::evdns_base_new,
+              decltype(evdns_base_free) destruct = ::evdns_base_free>
+    static Var<EvdnsBase> create(Var<EventBase> base,
+                                 bool initialize_nameservers = true,
+                                 bool fail_requests = true) {
```

Comment:
We cannot actually afford to leave the user the choice of whether failing the requests or not when the `EvdnsBase` object is destroyed. Saying this because for us pending DNS requests are bound to allocated `std::function<>` objects, hence, in order to properly cleanup, we MUST fail pending requests on exit, such that we have a chance to `delete` the memory that we have allocated.

In light of this, I'd probably refactor this class to call `::event_base_free()` in the destructor rather than doing so inside the smart pointer. I guess the former is going to be much less surprising for a C++ developer. To do this, you should probably make the class `template class` and pass the `::evdns_base_free` thing as argument to such template.